src/mpl/include/mpl_spinlock.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#ifndef MPL_SPINLOCK_H_INCLUDED
+#define MPL_SPINLOCK_H_INCLUDED
+
+#include <assert.h>
+#include "mplconfig.h"
+#include "mpl_atomic.h"
+
+#ifndef MPL_USE_NO_ATOMIC_PRIMITIVES
+
+/* Let's use atomic-based spinlock. */
+typedef struct MPL_spinlock_t {
+    MPL_atomic_int_t val;
+} MPL_spinlock_t;
+
+#define MPL_SPINLOCK_INITIALIZER { MPL_ATOMIC_INT_T_INITIALIZER(0) }
+
+static void MPL_spinlock_lock(MPL_spinlock_t * lock)
+{
+    while (MPL_atomic_cas_int(&lock->val, 0, 1));
+}
+
+static int MPL_spinlock_trylock(MPL_spinlock_t * lock)
+{
+    /* Return 0 if the lock is successfully taken.  Otherwise, return non-zero.
+     * This trylock is strong since MPL_atomic_cas_int is atomically strong. */
+    return MPL_atomic_cas_int(&lock->val, 0, 1);
+}
+
+static void MPL_spinlock_unlock(MPL_spinlock_t * lock)
+{
+    assert(MPL_atomic_relaxed_load_int(&lock->val) == 1);
+    MPL_atomic_release_store_int(&lock->val, 0);
+}
+
+#elif defined(MPL_HAVE_PTHREAD_H)
+
+/* If pthread can be used, let's use pthread_mutex. Note that we cannot use
+ * pthread's spinlock because it does not have a static initializer. */
+#include <pthread.h>
+
+#define MPL_spinlock_t pthread_mutex_t
+
+#define MPL_SPINLOCK_INITIALIZER PTHREAD_MUTEX_INITIALIZER
+
+static void MPL_spinlock_lock(MPL_spinlock_t * lock)
+{
+    int ret = pthread_mutex_lock(lock);
```

Comment:
I like the idea that we can swap the implementation. But that makes the name `SPIN` inappropriate. I am tempted to call it `INITLOCK`. In mpich we have `initshm`, `initcomm`, both refer to some facility that to be used before full initialization. What do you think?

Reply:
As an Argobots developer, this mutex implementation behaves like a spinlock, so this is spinlock.

I'm fine to call it `initlock` or whatever, but then does it become providing a static initializer for a mutex? If so, it needs extra efforts.

Reply:
> but then does it become providing a static initializer for a mutex?

I am not sure the above is syntactically complete, but yes, a static initializer is the only requirement for this lock

Reply:
@hzhou So are you suggesting that dropping this PR and providing a static initializer for `MPL_thread_mutex` instead?

Reply:
Thanks. I create `MPL_initlock`. To change the branch name, I will close this PR.